broadinstitute/consent
# [DUOS-1972][risk=medium] V3 dataset registration endpoint stub

## Addresses
Partially addresses https://broadworkbench.atlassian.net/browse/DUOS-1972

Marked as **Medium** risk due to the optional file upload.

New endpoint for registering a dataset submission. API takes a specific schema and generates N (minimum 1) datasets from the provided data. 

### This PR covers:
* API Endpoint
* API Documentation
  * Includes some minor category renaming and v2 deprecation
* Tests
* Schema validation

Future PRs will cover the creation of files and datasets from the provided schema.

All classes in `org.broadinstitute.consent.http.models.dataset_registration_v1` are generated via [jsonschema2pojo](https://github.com/joelittlejohn/jsonschema2pojo). These required some manual tweaks to work correctly with our classes, in particular `Double`s need to be `Integer`s and enums had case sensitivity issues.

### Example swagger

<img width="1261" alt="Screen Shot 2022-09-29 at 4 00 51 PM" src="https://user-images.githubusercontent.com/116679/193130436-60a968ab-bd49-44ac-9222-a69ebda2b76d.png">

### Testing

- Spin up the branch
- Construct sample registration
- Authenticate as admin or data submitter
- Post with and without file contents
- Putting the debugger on the endpoint should show that all content is being parsed and validated

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment
